src/coreclr/jit/compiler.hpp

```diff
@@ -4970,6 +4970,61 @@ unsigned Compiler::fgRunDfs(VisitPreorder visitPreorder, VisitPostorder visitPos
     return preOrderIndex;
 }
 
+//------------------------------------------------------------------------
+// fgVisitBlocksInLoopAwareRPO: Visit the blocks in 'dfsTree' in reverse post-order,
+// but ensure loop bodies are visited before loop successors.
+//
+// Type parameters:
+//   TFunc - Callback functor type
+//
+// Parameters:
+//   dfsTree     - The DFS tree of the flow graph
+//   blockToLoop - A mapping of each block in 'dfsTree' to its containing loop, if applicable
+//   func        - Callback functor that operates on a BasicBlock*
+//
+// Returns:
+//   A postorder traversal with compact loop bodies.
+//
+template <typename TFunc>
+void Compiler::fgVisitBlocksInLoopAwareRPO(FlowGraphDfsTree* dfsTree, BlockToNaturalLoopMap* blockToLoop, TFunc func)
+{
+    assert(dfsTree != nullptr);
+    assert(blockToLoop != nullptr);
+
+    EnsureBasicBlockEpoch();
+    BlockSet visitedBlocks(BlockSetOps::MakeEmpty(this));
+
+    auto visitBlock = [this, func, &visitedBlocks](BasicBlock* block) -> BasicBlockVisit {
+        // If this block is in a loop, we will try to visit it more than once
+        // (first when we visit its containing loop, and then later as we iterate
+        // through the initial RPO).
+        // Thus, we need to keep track of visited blocks.
+        if (!BlockSetOps::IsMember(this, visitedBlocks, block->bbNum))
+        {
+            func(block);
+            BlockSetOps::AddElemD(this, visitedBlocks, block->bbNum);
+        }
+
+        return BasicBlockVisit::Continue;
+    };
+
+    for (unsigned i = dfsTree->GetPostOrderCount(); i != 0; i--)
+    {
+        BasicBlock* const           block = dfsTree->GetPostOrder(i - 1);
+        FlowGraphNaturalLoop* const loop  = blockToLoop->GetLoop(block);
+
+        // If this block is a loop header, visit the entire loop before moving on
+        if ((loop != nullptr) && (block == loop->GetHeader()))
+        {
+            loop->VisitLoopBlocksReversePostOrder(visitBlock);
+        }
```

Comment:
I don't think this handles nested loops properly. This needs some form of recursion, probably -- similarly to `fgValueNumberBlocks`.

I think this utility can be implemented without a dependency on `BlockToNaturalLoopMap` since `FlowGraphNaturalLoops` stores loops in descendant order of the header's post order number, so `FlowGraphNaturalLoops::GetLoopByHeader` can have an efficient binary search implementation (there is a TODO about it). It should also be possible to walk the current loop and current block in lockstep, although it seems unnecessary to go that far.

Reply:
> I think this utility can be implemented without a dependency on BlockToNaturalLoopMap since FlowGraphNaturalLoops stores loops in descendant order of the header's post order number

That seems simple enough -- I'll change it.

> This needs some form of recursion, probably -- similarly to fgValueNumberBlocks.

I'm guessing we can't use a lambda for the recursive logic, right? If we need to split the recursive logic into another method, would it make sense to put the loop-aware RPO logic in some sort of visitor class to hide the recursive details?